connector/dao/src/main/java/it/pagopa/selfcare/mscore/connector/dao/UserConnectorImpl.java

```diff
@@ -163,21 +165,46 @@ public void findAndUpdateState(String userId, @Nullable String relationshipId, @
     }
 
     @Override
-    public void findAndUpdateStateByInstitutionAndProduct(String userId, String institutionId, String productId, RelationshipState state) {
-        Query query = Query.query(Criteria.where(UserEntity.Fields.id.name()).is(userId));
+    public void findAndUpdateStateWithOptionalFilter(String userId, String institutionId, String productId, PartyRole role, String productRole, RelationshipState state) {
 
-        Update update = new Update()
-                .set(constructQuery(CURRENT_USER_BINDING_REF, UserBinding.Fields.products.name(), CURRENT_PRODUCT_REF, OnboardedProduct.Fields.status.name()), state)
-                .set(constructQuery(CURRENT_USER_BINDING_REF, UserBinding.Fields.products.name(), CURRENT_PRODUCT_REF, OnboardedProduct.Fields.updatedAt.name()), OffsetDateTime.now());
+        checkRoles(productId, role, productRole);
 
-        update.filterArray(Criteria.where(CURRENT_PRODUCT + OnboardedProduct.Fields.productId.name()).is(productId)
-                .and(CURRENT_PRODUCT + OnboardedProduct.Fields.status.name()).is(RelationshipState.ACTIVE.name()));
-        update.filterArray(Criteria.where(CURRENT_USER_BINDING + UserBinding.Fields.institutionId.name()).is(institutionId));
+        Query query = Query.query(Criteria.where(UserEntity.Fields.id.name()).is(userId));
+        Update update = new Update();
+
+        boolean productFilterIsPresent = role != null || StringUtils.hasText(productId) || StringUtils.hasText(productRole);
+
+        if (StringUtils.hasText(institutionId) && productFilterIsPresent) {
```

Comment:
This is not needed for this PR, just an eventual hint.

Fot this kind of ops, another approach is let the caller to set criteria chosing from a restricted set of filters.

Something alike:

```java
public interface UserFilterCriteria {
    Criteria toCriteria();
}

public class InstitutionFilterCriteria implements UserFilterCriteria {
    private String institutionId;

    public InstitutionFilterCriteria(String institutionId) {
        this.institutionId = institutionId;
    }

    @Override
    public Criteria toCriteria() {
        return Criteria.where(CURRENT_USER_BINDING + UserBinding.Fields.institutionId.name()).is(institutionId);
    }
}

public void findAndUpdateStateWithOptionalFilter(List<UserFilterCriteria> criteriaList, RelationshipState state) {
    Query query = new Query();
    for (FilterCriteria filterCriteria : criteriaList) {
        query.addCriteria(filterCriteria.toCriteria());
    }

    // ...

    repository.findAndModify(query, update, options, UserEntity.class);
}

// caller
List<FilterCriteria> criteriaList = new ArrayList<>();
criteriaList.add(new InstitutionFilterCriteria(institutionId));
// Add other criteria...

findAndUpdateStateWithOptionalFilter(criteriaList, someRelationshipState);
```

In this way you don't have to modify and adapt the logic inside `findAndUpdateStateWithOptionalFilter` when adding new criteria and you avoid to add new arguments which are global to the whole method, but limited to one specific criteria.